File: apps/flutter_wallet/lib/src/core/native/parsers/wallets.dart
import '../../../models.dart';
import '../../../wallet_api.dart';
import 'fields.dart';

WalletSummary walletFromNativeJson(Object? value) {
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return WalletSummary(
    id: nativeStringField(value, 'id'),
    label: nativeStringField(value, 'label'),
  );
}

AccountSummary accountFromNativeJson(Object? value) {
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return AccountSummary(
    walletId: nativeStringField(value, 'wallet_id'),
    chain: nativeStringField(value, 'chain'),
    address: nativeStringField(value, 'address'),
    derivationPath: nativeStringField(value, 'derivation_path'),
  );
}

KeystoreExport keystoreExportFromNativeJson(Object? value) {
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return KeystoreExport(
    walletId: nativeStringField(value, 'wallet_id'),
    label: nativeStringField(value, 'label'),
    secretKind: nativeStringField(value, 'secret_kind'),
    ciphertextB64: nativeStringField(value, 'ciphertext_b64'),
    nonceB64: nativeStringField(value, 'nonce_b64'),
    saltB64: nativeStringField(value, 'salt_b64'),
    kdfName: nativeStringField(value, 'kdf_name'),
    kdfParamsJson: nativeStringField(value, 'kdf_params_json'),
    cipherName: nativeStringField(value, 'cipher_name'),
    version: nativeIntField(value, 'version'),
  );
}

File: crates/wallet_core/src/models/wallets.rs
use super::ChainId;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WalletSummary {
    pub id: Uuid,
    pub label: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct KeystoreExport {
    pub wallet_id: Uuid,
    pub label: String,
    pub secret_kind: String,
    pub ciphertext_b64: String,
    pub nonce_b64: String,
    pub salt_b64: String,
    pub kdf_name: String,
    pub kdf_params_json: String,
    pub cipher_name: String,
    pub version: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Account {
    pub wallet_id: Uuid,
    pub chain: ChainId,
    pub address: String,
    pub derivation_path: String,
    pub account_index: u32,
}
